src/com/google/cloud/datastore/node/DatastoreDAO.js

```diff
@@ -0,0 +1,267 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'com.google.cloud.datastore.node',
+  name: 'SelectData',
+
+  documentation: function() {/*
+                               State passed around by intermediate callbacks
+                               during a select() in progress. These data must
+                               be retained to notify the sink, send the
+                               correct payload to subsequent API calls, and
+                               return results in the Promise.
+                              */},
+
+  properties: [
+    {
+      name: 'sink'
+    },
+    {
+      name: 'requestPayload'
+    },
+    {
+      class: 'Array',
+      name: 'results',
+    }
+  ]
+});
+
+foam.CLASS({
+  package: 'com.google.cloud.datastore.node',
+  name: 'DatastoreDAO',
+  extends: 'foam.dao.AbstractDAO',
+
+  documentation: function() {/*
+                               DAO implementation that speaks of Cloud
+                               Datastore V1 REST API.
+
+                               https://cloud.google.com/datastore/docs/reference/rest/
+
+                               This implementation uses structured queries,
+                               not GQL queries.
+                              */},
+
+  requires: [
+    'com.google.cloud.datastore.node.SelectData',
+    'foam.dao.ArraySink',
+    'foam.net.universal.HTTPRequest'
+  ],
+  imports: [ 'projectId' ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'protocol',
+      value: 'https',
+      final: true
+    },
+    {
+      class: 'String',
+      name: 'host',
+      value: 'datastore.googleapis.com',
+      final: true
+    },
+    {
+      class: 'Int',
+      name: 'port',
+      value: 443,
+      final: true
+    },
+    {
+      class: 'String',
+      name: 'baseURL',
+      factory: function() {
+        return this.protocol + '://' + this.host + ':' + this.port +
+            '/v1/projects/' + this.projectId;
+      }
+    }
+  ],
+
+  methods: [
+    function getRequest(name, payload) {
+      var headers = { Accept: 'application/json' };
+      if ( payload ) headers['Content-Type'] = 'application/json';
+      return this.HTTPRequest.create({
+        method: 'POST',
+        url: this.baseURL + ':' + name,
+        headers: headers,
+        responseType: 'json',
+        payload: payload
+      });
+    },
+
+    function find(id) {
+      var key = foam.core.FObject.isInstance(id) ?
+          id.getDatastoreKey() : this.getDatastoreKeyFromId(id);
+      return this.getRequest('lookup', JSON.stringify({ keys: [ key ] })).send()
```

Comment:
This isn't operating over FOAM objects. Is JSON.Compact.stringify() faster than JSON.stringify()? If not, I think this is the right strategy there.